### PR TITLE
SPARK-6045 RecordWriter should be checked against null in PairRDDFunctio...

### PR DESCRIPTION
**Pull request by tedyu**
...ns#saveAsNewAPIHadoopDataset

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 1]
```diff
@@ -1004,7 +1004,9 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
           recordsWritten += 1
         }
       } finally {
-        writer.close(hadoopContext)
+        if (writer != null) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Why check here? if it's `null` then this fails already in the loop. Just assert it's not `null` with `require` after it's obtained, if this is supposed to die in this case anyway.

Comment (tedyu) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Thanks for the review, updated as suggested.

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 2]
```diff
@@ -993,6 +993,7 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
       val (outputMetrics, bytesWrittenCallback) = initHadoopOutputMetrics(context)
 
       val writer = format.getRecordWriter(hadoopContext).asInstanceOf[NewRecordWriter[K,V]]
+      require(writer != null, s"Unable to obtain NewRecordWriter")
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: There's no string interpolation here, and there is no actual `NewRecordWriter` class (it's an alias). Don't worry I can edit that before merge. I will have Jenkins test just for completeness

**Comment by srowen** [comment]
OK to test

**Comment by srowen** [comment]
Jenkins, test this please

**Comment by srowen** [comment]
OK this looks like no-risk small fix. I'll adjust it and merge.

**Pull request closed without merging**